src/extra/gpkg/read.jl

```diff
@@ -0,0 +1,482 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+
+# Main entry point
+function gpkgread(fname::String; layer::Int=1, kwargs...)
+  db = SQLite.DB(fname)
+  
+  # 1. Read metadata to extract names of tables and CRS type
+  table_name, geom_column, crs_type = gpkgmetadata(db, layer)
+  
+  # 2. Get all column names from the table
+  columns_result = DBInterface.execute(db, "PRAGMA table_info(\"$table_name\")")
+  all_columns = [row.name for row in Tables.rows(columns_result)]
+  
+  # Filter out geometry column to get only attribute columns
+  attr_columns = filter(col -> col != geom_column, all_columns)
+  
+  # 3. Execute queries to get attribute and geometry data separately
+  if !isempty(attr_columns)
+    # Build query for attribute columns only
+    attr_columns_str = join(["\"$col\"" for col in attr_columns], ", ")
+    table_result = DBInterface.execute(db, "SELECT $attr_columns_str FROM \"$table_name\"")
+    # Materialize the query result into a NamedTuple
+    cols = Tables.columns(table_result)
+    names = Tables.columnnames(cols)
+    table = NamedTuple{names}([
+      Tables.getcolumn(cols, name) for name in names
+    ])
+  else
+    # No attribute columns, only geometry
+    table = nothing
+  end
+  
+  geom_result = DBInterface.execute(db, "SELECT \"$geom_column\" FROM \"$table_name\"")
+  
+  # 4. Read geometries (including missing/nothing values for asgeotable to handle)
+  geoms = gpkgreadgeoms(geom_result, geom_column, crs_type)
+  
+  close(db)
+  
+  # 5. Combine attributes with geoms into a Tables.jl table
+  # asgeotable() will handle missing/nothing geometries automatically
+  if isnothing(table)
+    # Only geometry column
+    return (geometry=geoms,)
+  else
+    # Combine attribute columns with geometry column
+    cols = Tables.columns(table)
+    names = Tables.columnnames(cols)
+    combined_cols = NamedTuple{(names..., :geometry)}([
+      [Tables.getcolumn(cols, name) for name in names]...,
+      geoms
+    ])
+    return combined_cols
+  end
+end
+
+function gpkgmetadata(db::SQLite.DB, layer::Int)
+  # Get feature table names
+  contents_result = DBInterface.execute(db,
+    """
+    SELECT table_name, data_type
+    FROM gpkg_contents
+    WHERE data_type = 'features'
+    ORDER BY table_name
+    """)
+  
+  # Use Iterators.drop to get the requested layer
+  rows = Tables.rows(contents_result)
+  layer_iter = Iterators.drop(rows, layer - 1)
+  
+  # Try to get the first row
+  first_row = nothing
+  try
+    first_row = first(layer_iter)
+  catch
+    error("Layer $layer not found in GeoPackage")
+  end
```

Comment:
Sorry, I believe you used `iterate` to avoid the `try-catch`? Please stick to `iterate` if `first` throws on empty iterators.